### PR TITLE
fix(cacao): reject expired and not-before timestamps

### PR DESCRIPTION
**Pull request by SashaMIT**
## Summary

`CacaoObject.VerifySignature` checked the signature only. It formatted `exp` and `nbf` into the CAIP-122 message, then returned true for any valid signature, including after `exp`, before `nbf`, or when those fields were unparseable.

Session authenticate (`Engine`, `EngineHandler`) and SIWE (`SiweController.VerifyMessageAsync`) treat that boolean as the gate. Request-level `ExpiryTimestamp` is a different field.

Fail closed when `exp` is present and past or unparseable, and when `nbf` is present and future or unparseable. Absent timestamps stay optional, matching CAIP-122 / EIP-4361.

Language-split of WalletConnect/walletconnect-monorepo#7312, reown-com/reown-kotlin#427, and reown-com/reown-rust#109. Distinct from reown-swift#345 (CR/LF in SIWE fields).

**Threat-model:** the attacker already holds a previously valid signed CACAO. They do not control the verifier clock or the key. After `exp` (or before `nbf`) the signature is still valid, and `VerifySignature` still returned true. That is replay after expiry, not host or agent authority.

## Test plan

- [x] `dotnet test test/Reown.Sign.Test/Reown.Sign.Test.csproj --filter FullyQualifiedName~CacaoTests` (12/12, net8.0)
- [x] `dotnet test test/Reown.Sign.Test/Reown.Sign.Test.csproj --filter Category=unit` (206/206, net8.0)
- [x] Revert-tested: removing the `IsWithinValidityWindow` check in `CacaoObject.VerifySignature` makes `VerifySignature_ExpiredCacao_ReturnsFalseWithoutThrowing` fail (empty signature reaches `SignatureUtils` and throws)

### CODE REVIEW

**Comment by SashaMIT** [comment]
I have read the CTA Document and I hereby sign the CTA

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs [hunk 1]
```diff
@@ -27,6 +27,11 @@ public CacaoObject(CacaoHeader header, CacaoPayload payload, CacaoSignature sign
 
         public async Task<bool> VerifySignature(string projectId, string rpcUrl = null)
         {
+            if (!Payload.IsWithinValidityWindow())
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs]: This makes a pre-existing ambiguity fatal: `Engine.cs:955` copies `AuthParams.Expiration` verbatim into the payload `exp`, while line 971 parses the same string as a duration in seconds. A consumer passing `"3600"` ends up with `exp: "3600"`, which doesn't parse — so a previously working flow now fails closed after the user has signed. Could you normalize in `Engine.cs`: when `Expiration` parses as a long, emit `DateTimeOffset.UtcNow.AddSeconds(n).ToRfc3339()` (same as `iat` a line above)?

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs [hunk 2]
```diff
@@ -27,6 +27,11 @@ public CacaoObject(CacaoHeader header, CacaoPayload payload, CacaoSignature sign
...
+            {
+                return false;
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs:34]: Both Engine call sites will report this as "Invalid cacao signature", which points debugging at the wrong cause. Could you log the actual reason (expired / not yet valid / unparseable) before returning false?

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs [hunk 3]
```diff
@@ -68,6 +69,42 @@ public CacaoPayload(
...
 
+        /// <summary>
+        ///     Returns true when optional CACAO <c>exp</c> and <c>nbf</c> are absent or <paramref name="now"/>
+        ///     is inside the validity window. Unparseable timestamps fail closed.
+        /// </summary>
+        public bool IsWithinValidityWindow(DateTimeOffset? now = null)
+        {
+            var clock = now ?? DateTimeOffset.UtcNow;
+
+            if (Expiration != null)
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs]: `!= null` treats `"exp": ""` (from a peer that always emits the key) as present, so it fails to parse and gets rejected. `!string.IsNullOrWhiteSpace` here and for `nbf` would treat blank values as absent.

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs [hunk 4]
```diff
@@ -68,6 +69,42 @@ public CacaoPayload(
...
+                if (!TryParseCacaoTimestamp(NotBefore, out var nbf) || clock < nbf)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool TryParseCacaoTimestamp(string value, out DateTimeOffset timestamp)
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs:132]: Docs nit: the summary above references `<paramref name="now"/>` without a `<param>` tag, and this method has no XML docs — the repo asks for them on new methods.

File: test/Reown.Sign.Test/CacaoTests.cs [hunk 5]
```diff
@@ -48,4 +64,88 @@ public void FormatMessage_WithoutRecap_ReturnsExpectedMessage()
...
+    [Fact] [Trait("Category", "unit")]
+    public void IsWithinValidityWindow_PastNbf_IsValid()
+    {
+        Assert.True(Payload(nbf: "2023-01-01T00:00:00Z").IsWithinValidityWindow(Now));
+    }
+
+    [Fact] [Trait("Category", "unit")]
+    public void IsWithinValidityWindow_FutureNbf_IsNotYetValid()
+    {
+        Assert.False(Payload(nbf: "2024-01-01T00:00:00Z").IsWithinValidityWindow(Now));
```
Comment (skibitsky) [inline test/Reown.Sign.Test/CacaoTests.cs:102]: Could you add the `nbf == now` boundary case? Flipping `clock < nbf` to `<=` would keep all current tests green while wrongly rejecting a valid CACAO. A timestamp with a non-UTC offset (e.g. `+02:00`) is also worth pinning.

**Review by skibitsky — CHANGES_REQUESTED** [review]
Hey @SashaMIT, thanks for this, great first contribution! 👏

One thing to resolve before merge (inline on CacaoObject.cs): Engine.cs uses `AuthParams.Expiration` both as a timestamp and as a duration in seconds, and the new check turns that ambiguity into a runtime failure for consumers who follow the seconds reading. Normalizing it at the source keeps your fail-closed behavior intact without breaking anyone.

The other inline comments are small: log the rejection reason so expiry isn't reported as a bad signature, treat empty exp/nbf as absent, add the nbf == now boundary test, and an XML docs nit.

Heads-up on CI: Unity job failures are on our side, not your change.

Looking forward to the next revision — thanks again!

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs (hunk 1 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs]: Done. Engine now runs Expiration through NormalizeExpiration: an integer becomes UtcNow plus that many seconds as RFC 3339, same clock as iat.

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs (hunk 2 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Models/Cacao/CacaoObject.cs:34]: VerifySignature now logs CACAO rejected: expired / not yet valid / unparseable exp|nbf before returning false.

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs (hunk 3 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs]: Switched both checks to IsNullOrWhiteSpace so a blank exp or nbf is treated as absent.

File: src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs (hunk 4 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Models/Cacao/CacaoPayload.cs:132]: Added the param tag and XML docs on the window check plus TryParseCacaoTimestamp.

File: test/Reown.Sign.Test/CacaoTests.cs (hunk 5 above)
Comment (SashaMIT) [reply test/Reown.Sign.Test/CacaoTests.cs:102]: Added nbf == now (valid), a +02:00 exp pair, and blank exp/nbf as absent. Left clock < nbf as-is.

**Comment by SashaMIT** [comment]
Hey, thanks for the review and the welcome. That Engine ambiguity is a real one. A consumer passing 3600 as seconds would have signed a CACAO that then failed closed.

Engine now normalizes an integer Expiration to UtcNow plus that many seconds as RFC 3339, same as iat. The window check stays fail-closed. VerifySignature logs the actual reason (expired / not yet valid / unparseable) before returning false. Blank exp/nbf are treated as absent. Added the nbf == now case, a non-UTC offset pin, and XML docs.

Pushed on 7bc25d5. CacaoTests 18/18 on net8.0. Unity reds are still yours, as you said.

File: src/Reown.Sign/Runtime/Utils/CacaoUtils.cs [hunk 6]
```diff
@@ -47,5 +47,28 @@ public static string ToRfc3339(this DateTimeOffset dateTimeOffset)
...
+        public static string NormalizeExpiration(string expiration, DateTimeOffset? now = null)
+        {
+            if (string.IsNullOrWhiteSpace(expiration))
+            {
+                return expiration;
+            }
+
+            if (long.TryParse(expiration, NumberStyles.Integer, CultureInfo.InvariantCulture, out var seconds))
+            {
+                return (now ?? DateTimeOffset.UtcNow).AddSeconds(seconds).ToRfc3339();
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Utils/CacaoUtils.cs]: Two integer inputs slip past this: epoch milliseconds (`"1735689600000"`) overflow `AddSeconds` and throw `ArgumentOutOfRangeException` uncaught at the call site, and epoch seconds (`"1735689600"`, a dapp intending 2025-01-01) expand to `now + 55 years` — the expiry check fails open for that consumer. A range guard fixes both: only expand positive values below a sane cap (e.g. one year) and let the rest fall through unchanged, so they fail closed as unparseable. While here, add `.ToUniversalTime()` before `AddSeconds` — `ToRfc3339` stamps a literal `Z` without converting, so a non-UTC `now` would mislabel the instant.

File: src/Reown.Sign/Runtime/Engine.cs [hunk 7]
```diff
@@ -952,7 +952,7 @@ await Task.WhenAll(
                 Version = "1",
                 Nonce = authParams.Nonce,
                 Iat = DateTimeOffset.UtcNow.ToRfc3339(),
-                Exp = authParams.Expiration?.ToString(),
+                Exp = CacaoUtils.NormalizeExpiration(authParams.Expiration),
                 Nbf = authParams.NotBefore?.ToString(),
```
Comment (skibitsky) [inline src/Reown.Sign/Runtime/Engine.cs]: `Nbf` is still passed through raw, so a consumer who passes seconds here (like `Expiration` now accepts) gets rejected as "unparseable nbf" at approve time. Either normalize it the same way or document the accepted formats on `AuthParams.Expiration`/`NotBefore` — those fields have no XML docs, which is where the original ambiguity came from.

File: test/Reown.Sign.Test/CacaoTests.cs [hunk 8]
```diff
@@ -48,4 +65,129 @@ public void FormatMessage_WithoutRecap_ReturnsExpectedMessage()
...
+        Assert.False(Payload(exp: "2023-01-01T00:00:00Z").IsWithinValidityWindow(Now, out var reason));
+        Assert.Equal("expired", reason);
+    }
+
+    [Fact] [Trait("Category", "unit")]
+    public void NormalizeExpiration_IntegerSeconds_BecomesRfc3339()
+    {
+        var now = DateTimeOffset.FromUnixTimeSeconds(1_700_000_000);
+        var normalized = CacaoUtils.NormalizeExpiration("3600", now);
+        Assert.Equal(now.AddSeconds(3600).ToRfc3339(), normalized);
```
Comment (skibitsky) [inline test/Reown.Sign.Test/CacaoTests.cs]: The expectation re-runs the production expression, so a wrong format string can't fail — could you assert the literal `"2023-11-14T23:13:20.0000000Z"`? A large (epoch-like) and a negative integer case would also pin the range behavior.

**Review by skibitsky — COMMENTED** [review]
Thanks for the quick turnaround @SashaMIT, all five earlier comments are addressed and the threads are resolved. The boundary tests and the reason logging came out great 🙌

One more round: the new `NormalizeExpiration` treats any integer as a duration, which throws uncaught on epoch-milliseconds input and quietly expands epoch-seconds into a decades-long validity window, reopening the hole this PR closes. A small range guard fixes both. Details inline, along with two smaller notes on `nbf` docs and a test expectation.

Really close now, thanks again for sticking with it!

File: src/Reown.Sign/Runtime/Utils/CacaoUtils.cs (hunk 6 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Utils/CacaoUtils.cs]: Done. Only positive values at or below Clock.ONE_YEAR expand. Epoch milliseconds, epoch seconds, and negatives stay unchanged so they fail closed instead of overflowing or opening a decades-long window. ToUniversalTime runs before AddSeconds so a non-UTC now still stamps Z.

File: src/Reown.Sign/Runtime/Engine.cs (hunk 7 above)
Comment (SashaMIT) [reply src/Reown.Sign/Runtime/Engine.cs]: Nbf now goes through the same helper. AuthParams.Expiration and NotBefore have XML docs for RFC 3339 vs duration-in-seconds.

File: test/Reown.Sign.Test/CacaoTests.cs (hunk 8 above)
Comment (SashaMIT) [reply test/Reown.Sign.Test/CacaoTests.cs]: Asserts the literal 2023-11-14T23:13:20.0000000Z. Added epoch-like, negative, and millisecond cases, plus a +02:00 now pin.

**Comment by SashaMIT** [comment]
Thanks for catching that, the epoch-seconds expansion was a real hole and would have reopened what this PR closes.

Range guard now only expands positive values at or below Clock.ONE_YEAR. Epoch ms, epoch seconds, and negatives stay as-is and fail closed. ToUniversalTime before AddSeconds so a non-UTC now still emits Z. Nbf uses the same helper, and AuthParams.Expiration / NotBefore now document the accepted formats.

Tests assert the literal timestamp plus the large and negative cases. Pushed on 53b26a5. CacaoTests 22/22 on net8.0.

**Review by skibitsky — APPROVED** [review]
